sound/soc/sof/ipc4-topology.c

```diff
@@ -966,6 +966,14 @@ sof_ipc4_update_resource_usage(struct snd_sof_dev *sdev, struct snd_sof_widget *
 	pipe_widget = swidget->spipe->pipe_widget;
 	pipeline = pipe_widget->private;
 	pipeline->mem_usage += total;
+
+	/* Update base_config->cpc from the module manifest */
+	sof_ipc4_update_cpc_from_manifest(sdev, swidget->module_info, base_config);
+
+	dev_dbg(sdev->dev, "%s: ibs / obs / cpc: %u / %u / %u\n",
+		swidget->widget->name, base_config->ibs, base_config->obs,
+		base_config->cpc);
```

Comment:
... so here are logging information that is inconsistent because there is no matching ibs/obs/cpc triplet.

Reply:
Line 478 prints the information if there is no matching IBS/OBS was found, but we did ended up with a non 0 CPC.
If you have deb level enabled it will look like this:
```
COPIER (UUID: 9BA00C83-CA12-4A83-943C-1FA2E82F9DDA): No CPC match for ibs/obs: 192/384
copier.host.1.1: ibs / obs / cpc: 192 / 384 / 7272
```
It is not relevant in the dev_info() print what is the CPC we ended up using, what it matters is that we have no match.

If there is no match, we will print with dev_warn() first:
```
COPIER (UUID: 9BA00C83-CA12-4A83-943C-1FA2E82F9DDA): No CPC value available
copier.host.1.1: ibs / obs / cpc: 192 / 384 / 0
```

Reply:
@ujfalusi I am not clear why there are three levels of information with warn, info and dbg. Two of them are errors in my book.

Reply:
and have no audio card at all?
It might take 4 months for a user to get a new firmware with updated manifest..

What if someone experimenting with topology files and starts to use a module which does not happen to have the exact same config measured? We break audio?

If we do that with this PR than the next step is a revert of the series as it introduces a regression for users.

I don't think that is something we can do and should do.

All red cavs and avs PR test would be granted.

Reply:
If you don't have dev_dbg() enabled then the information about the use ibs/obs/cpc is not shown. User will see either a warning that there were no CPC at all (the ibs/obs is irrelevant as there is not a single one can be found in mod_cfg) or an info print that while there are CPC values, the ibs/obs pair has no matching hit.

And if I think about this, then likely we need to warn if there is no mod_cfg either for the module from where we would be taking the CPC.

Reply:
We are asked to parse the manifest and provide the CPC value for a given IBS/OBS. If the CPC doesn't exist for this configuration, well end of story. The driver should not invent information that should have been provided by the firmware integrator.

If someone tinkers with topology file and breaks the CPC mapping, it's a revert of the topology files. very simple. If they ship a broken topology without testing that it is compatible with the manifest, there's nothing we can do.

We have lots of cases like this today, if someone changes the topology and refers to a non-existing firmware component, then we just don't have a card.

We just don't have ways to deal with bad topology configurations other than failing to create a card.

Reply:
So, you want to fail with error and not allow the playback if we  don't have CPC with direct match from the manifest?
Either if  the mod_cfg does not have a matching entry or if there is no mod_cfg at all?

If we do this than you are not going to be able to bisect firmware beyond the point the manifest has been updated as the kernel will refuse to work.

Reply:
if the kernel worked in the past and suddendly refuses to work, that would be because of a topology change that modified IBS/OBS, or an update of the manifest. At that point, the changed configuration will be detected as failing CI tests and prevent the update from being merged. 

If we don't fail, we will let a bad configuration be merged and no one will ever attempt to fix it. It's the responsibility of the firmware team to make sure the manifest makes sense and covers all allowed usages.